python_modules/libraries/dagster-aws/dagster_aws/pipes/clients/emr_serverless.py

```diff
@@ -0,0 +1,289 @@
+
+        if isinstance(self.context_injector, PipesEnvContextInjector):
+            dagster_env_vars.update(session.get_bootstrap_env_vars())
+
+        if "jobDriver" not in params:
+            params["jobDriver"] = {}
+
+        if "sparkSubmit" not in params["jobDriver"]:
+            params["jobDriver"]["sparkSubmit"] = {}
+
+        params["jobDriver"]["sparkSubmit"]["sparkSubmitParameters"] = params.get(
+            "jobDriver", {}
+        ).get("sparkSubmit", {}).get("sparkSubmitParameters", "") + "".join(
+            [
+                f" --conf spark.emr-serverless.driverEnv.{key}={value}"
+                for key, value in dagster_env_vars.items()
+            ]
+        )
+
+        return params
+
+    def _start(
+        self, context: OpExecutionContext, params: "StartJobRunRequestRequestTypeDef"
+    ) -> "StartJobRunResponseTypeDef":
+        response = self._client.start_job_run(**params)
+        job_run_id = response["jobRunId"]
+        context.log.info(
+            f"[pipes] {self.AWS_SERVICE_NAME} job started with job_run_id {job_run_id}"
+        )
+        return response
+
+    def _wait_for_completion(
+        self, context: OpExecutionContext, start_response: "StartJobRunResponseTypeDef"
+    ) -> "GetJobRunResponseTypeDef":
+        job_run_id = start_response["jobRunId"]
+
+        while response := self._client.get_job_run(
+            applicationId=start_response["applicationId"],
+            jobRunId=job_run_id,
+        ):
+            state: JobRunStateType = response["jobRun"]["state"]
+
+            if state in ["FAILED", "CANCELLED", "CANCELLING"]:
+                context.log.error(
+                    f"[pipes] {self.AWS_SERVICE_NAME} job {job_run_id} terminated with state: {state}. Details:\n{response['jobRun'].get('stateDetails')}"
+                )
+                raise RuntimeError(f"{self.AWS_SERVICE_NAME} job failed")
+
+            elif state == "SUCCESS":
+                context.log.info(
+                    f"[pipes] {self.AWS_SERVICE_NAME} job {job_run_id} completed with state: {state}"
+                )
+                return response
+
+            time.sleep(10)
+
+    def _read_messages(self, context: OpExecutionContext, response: "GetJobRunResponseTypeDef"):
+        application_id = response["jobRun"]["applicationId"]
+        job_id = response["jobRun"]["jobRunId"]
+
+        application = self.client.get_application(applicationId=application_id)["application"]
+
+        application_type = application["type"]
+
+        if application_type == "Spark":
+            worker_type = "SPARK_DRIVER"
+        elif application_type == "Hive":
+            worker_type = "HIVE_DRIVER"
+        else:
+            raise NotImplementedError(f"Application type {application_type} is not supported")
+
+        if isinstance(self.message_reader, PipesCloudWatchMessageReader):
+            # https://docs.aws.amazon.com/emr/latest/EMR-Serverless-UserGuide/logging.html#jobs-log-storage-cw
+
```

Comment:
minor nit.

I think doing 

```python
if not isinstance(self.message_reader, PipesCloudWatchMessageReader):
            context.log.warning(
                f"[pipes] {self.message_reader} is not supported for {self.AWS_SERVICE_NAME}. Dagster won't be able to receive logs and messages from the job."
            )
           return

# rest of it without indent
```

will increase readability quite a bit. This is a very "long" if block